ui/pages/editor.tsx

```diff
@@ -1,110 +1,254 @@
   const editorRef = useRef();
-  const [name, setName] = useState(props.fileName);
   const [fileTree, setFileTree] = useState<IFileTreeNode[]>([]);
-  const [codeRoute, setCodeRoute] = useState(initialCodeRoute)
+  const [monacoInstance, setMonacoInstance] = useState(null);
+  const [monacoInstanceLoaded, setMonacoInstanceLoaded] = useState(false);
+  const [editorModels, setEditorModels] = useState<Map<string, EditorModel>>(new Map<string, EditorModel>());
+  const [fileNodeMap, setFileNodeMap] = useState({});
+  const [newFileName, setNewFileName] = useState("");
+  const [selectedFileKey, setSelectedFileKey] = useState("");
 
   // https://github.com/vercel/swr/issues/284 - using initialData for SSR has open bug
-  const { data: code, error: codeError } = useSWR(name ? codeRoute : null, codeFetcher, { initialData: name === initialName ? initialCode : undefined }); 
+  // TODO: confirm this still works as expected
+  const { data: code, error: codeError, isValidating: codeValidating } = useSWR(selectedFileKey ? ['/static', selectedFileKey] : null, codeFetcher, { onSuccess: onCodeFetchSuccess})
+
+  async function codeFetcher(baseUrl, fileKey) {
+    const url = path.join(baseUrl, fileKey);
+    const data = await fetch(url + window.location.search);
+    const code = await data.text();
+    return code;
+  };
+
+  // when we have our code, fetch the proper model or create a new model to mount the code to
+  function onCodeFetchSuccess(data) {
+    let fileName = selectedFileKey;
+    if (!(fileName in editorModels)) {
+      const model = createEditorModel(fileName, data);
+      editorModels[fileName] = model;
+      setEditorModels({...editorModels});
+    }
+    editorRef.current.setModel(editorModels[fileName].model);
+  }
 
   useEffect(() => {
-      const transformedData = _transformFileTreeData(props.fileTree);
-      setFileTree(transformedData);
-      if(props.fileName) {
-        handleInitOpenFile(props.fileName);
-      }
+      const initializeMonaco = async () => {
+        const monacoInst = await monaco.init();
+        setMonacoInstance(monacoInst);
+        setMonacoInstanceLoaded(true);
+      };
+      initializeMonaco();
   }, []);
 
+  // Wait for editor setup before setting up the rest of the page data (easier to deal with that way)
+  useEffect(() => {
+    if(monacoInstanceLoaded && isEditorReady) {
+      initializeFileTreeData(props.fileTree);
+      selectFile(props.fileName);
+    }
+  }, [monacoInstanceLoaded, isEditorReady]);
+
   useEffect(() => {
     if (editorRef.current) {
-      bindSaveCommand(editorRef.current, () => saveFileAs(name));
+      bindSaveCommand(editorRef.current, () => saveFileAs(selectedFileKey));
     }
-    setCodeRoute(path.join('/code', name));
-  }, [name]);
+  }, [selectedFileKey]);
+
+  const selectFile = (fileName: string) => {
+    setSelectedFileKey(fileName); // should kick off process: new file key --> new code --> (new editor model) --> set editor model
```

Comment:
Basically order of operations: set a file key --> go get the code for that (useSWR) --> if we have an editor model loaded then use that...otherwise create an new editor model --> set the editor model onto the editor